core/src/main/scala/kafka/coordinator/transaction/TransactionCoordinator.scala

```diff
@@ -434,21 +435,49 @@ class TransactionCoordinator(brokerId: Int,
 
   private def abortTimedOutTransactions(): Unit = {
     txnManager.timedOutTransactions().foreach { txnIdAndPidEpoch =>
-      handleEndTransaction(txnIdAndPidEpoch.transactionalId,
-        txnIdAndPidEpoch.producerId,
-        txnIdAndPidEpoch.producerEpoch,
-        TransactionResult.ABORT,
-        (error: Errors) => error match {
-          case Errors.NONE =>
-            info(s"Completed rollback ongoing transaction of transactionalId: ${txnIdAndPidEpoch.transactionalId} due to timeout")
-          case Errors.INVALID_PRODUCER_ID_MAPPING |
-               Errors.INVALID_PRODUCER_EPOCH |
-               Errors.CONCURRENT_TRANSACTIONS =>
-            debug(s"Rolling back ongoing transaction of transactionalId: ${txnIdAndPidEpoch.transactionalId} has aborted due to ${error.exceptionName()}")
-          case e =>
-            warn(s"Rolling back ongoing transaction of transactionalId: ${txnIdAndPidEpoch.transactionalId} failed due to ${error.exceptionName()}")
-        })
-    }
+      val coordinatorEpochAndMetadata = txnManager.getTransactionState(txnIdAndPidEpoch.transactionalId).right.flatMap {
```

Comment:
Not sure, this is the pattern used everywhere else in this class for getting the transactionState. Perhaps @guozhangwang can comment?

Reply:
To close the loop here, we synced offline and the confusion is because of the semantics of the `right.flatMap` operation. the `getTransactionState` method returns `Either[Errors, Option[CoordinatorEpochoAndMetadata]]`. If it returns the `Left` the `right.flatMap` will still return `Left[Errors]`, which is what will be stored in `coordinatorEpochAndMetadata`. So we don't need to check it explicitly since that error will be returned directly in the subsequent code.